### PR TITLE
[SM-66] Create Secret Database Table

### PR DESCRIPTION
**Pull request by Thomas-Avery**
## Type of change

```
- [ ] Bug fix
- [X] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

The purpose of this PR is to create a database table, entity, and repository for the new Secret database table.

The new Secret table will use entity framework for all database providers. 

## Code changes

```.github/workflows/build.yml```
Updating so the "Build Swagger" step works with the new changes.

```bitwarden-server.sln```
Including references to newly created project.

```
bitwarden_license/src/Scim/packages.lock.json
bitwarden_license/src/Sso/packages.lock.json
bitwarden_license/test/CmmCore.Test/packages.lock.json
src/Admin/packages.lock.json
src/Api/packages.lock.json
src/Billing/packages.lock.json
src/Events/packages.lock.json
src/EventsProcessor/packages.lock.json
src/Icons/packages.lock.json
src/Identity/packages.lock.json
src/Infrastructure.EntityFramework/packages.lock.json
src/Notifications/packages.lock.json
src/SharedWeb/packages.lock.json
test/Api.Test/packages.lock.json
test/Billing.Test/packages.lock.json
test/Common/packages.lock.json
test/Core.Test/packages.lock.json
test/Icons.Test/packages.lock.json
test/Identity.IntegrationTest/packages.lock.json
test/Identity.Test/packages.lock.json
test/IntegrationTestCommon/packages.lock.json
util/MySqlMigrations/packages.lock.json
util/PostgresMigrations/packages.lock.json
util/SqlServerEFScaffold/packages.lock.json
src/Infrastructure.EntityFramework/Infrastructure.EntityFramework.csproj
```

Added the nuget package, Microsoft.EntityFrameworkCore.SqlServer versionn 6.0.4 to Infrastructure.EntityFramework, which triggered updates to all *.packages.lock.json files.

```
src/Core/Entities/Secret.cs
src/Infrastructure.EntityFramework/Models/Secret.cs
src/Core/Repositories/ISecretRepository.cs
src/Infrastructure.EntityFramework/Repositories/SecretRepository.cs
```
Entity Framework models, interface, and repository for the new Secret database table.
```
src/Infrastructure.EntityFramework/Repositories/DatabaseContext.cs
```
Entity Framework configuration for the new Secret table.

```
src/Infrastructure.EntityFramework/EntityFrameworkServiceCollectionExtensions.cs
src/SharedWeb/Utilities/ServiceCollectionExtensions.cs
```
Changes to allow the Secret repository to be in entity framework while all other repositories stay implemented in dapper when choosing the Microsoft SQL Server database provider.

```
src/Sql/dbo/Tables/Secret.sql
util/Migrator/DbScripts/2022-07-20_00_CreateSecretTable.sql
```
Microsoft SQL Server database code and migration script for the new Secret database table.

```
test/Core.Test/AutoFixture/EntityFrameworkRepositoryFixtures.cs
test/Core.Test/AutoFixture/SecretFixtures.cs
test/Core.Test/Helpers/Factories.cs
test/Core.Test/Repositories/EntityFramework/EqualityComparers/SecretCompare.cs
test/Core.Test/Repositories/EntityFramework/SecretRepositoryTests.cs
```
Adding in the ability for repository tests to use Microsoft SQL Server as a database provider. This is turned off by default.
Adding in the ability to determine which database provider is being used in a given unit test.
Adding secret repository tests.

```
util/MySqlMigrations/Migrations/20220720222516_CreateSecretTable.Designer.cs
util/MySqlMigrations/Migrations/20220720222516_CreateSecretTable.cs
util/MySqlMigrations/Migrations/DatabaseContextModelSnapshot.cs
util/MySqlMigrations/Scripts/2022-07-20_00_CreateSecretTable
```
MySQL migration changes for the new Secret database table.

```
util/PostgresMigrations/Factories.cs
```

With the upgrade to .NET 6.0 default timestamp behavior has changed to support legacy behavior for migrations, this application context must be set.

This is already set for Bitwarden at runtime here https://github.com/bitwarden/server/blob/7dfb04298d0b3d504a788b565bb0ff329302bb81/src/Infrastructure.EntityFramework/EntityFrameworkServiceCollectionExtensions.cs#L25-L27

```
util/PostgresMigrations/Migrations/20220720222109_CreateSecretTable.Designer.cs
util/PostgresMigrations/Migrations/20220720222109_CreateSecretTable.cs
util/PostgresMigrations/Migrations/DatabaseContextModelSnapshot.cs
util/PostgresMigrations/Scripts/2022-07-20_00_CreateSecretTable.psql
```
Postgres migration changes for the new Secret database table.

```
util/SqlServerEFScaffold/Factories.cs
util/SqlServerEFScaffold/SqlServerEFScaffold.csproj
util/SqlServerEFScaffold/README.MD
```

A new project created for engineers to validate their entity framework models with Microsoft SQL Server database changes.
Documentation on how to use is provided in the readme file.

## Before you submit

```
- [X] I have checked for formatting errors (`dotnet format --verify-no-changes`) (required)
- [X] If making database changes - I have also updated Entity Framework queries and/or migrations
- [X] I have added **unit tests** where it makes sense to do so (encouraged but not required)
- [ ] This change requires a **documentation update** (notify the documentation team)
- [ ] This change has particular **deployment requirements** (notify the DevOps team)
```

### CODE REVIEW

File: src/Infrastructure.EntityFramework/Repositories/DatabaseContext.cs [hunk 1]
```diff
@@ -90,13 +92,17 @@ protected override void OnModelCreating(ModelBuilder builder)
...
             eOrganizationApiKey.Property(c => c.Id).ValueGeneratedNever();
             eOrganizationConnection.Property(c => c.Id).ValueGeneratedNever();
+            eSecret.Property(c => c.Id).ValueGeneratedNever();
 
             eCollectionCipher.HasKey(cc => new { cc.CollectionId, cc.CipherId });
             eCollectionUser.HasKey(cu => new { cu.CollectionId, cu.OrganizationUserId });
             eCollectionGroup.HasKey(cg => new { cg.CollectionId, cg.GroupId });
             eGrant.HasKey(x => x.Key);
             eGroupUser.HasKey(gu => new { gu.GroupId, gu.OrganizationUserId });
+            eSecret.HasKey(s => s.Id).IsClustered();
```
Comment (Hinton) [inline src/Infrastructure.EntityFramework/Repositories/DatabaseContext.cs:103]:
```suggestion

            eSecret.HasKey(s => s.Id).IsClustered();
```

Comment (Hinton) [reply src/Infrastructure.EntityFramework/Repositories/DatabaseContext.cs:103]: I wonder if we should investigate using https://docs.microsoft.com/en-us/ef/core/modeling/#grouping-configuration to scale this file.

File: util/Migrator/DbScripts/2022-07-20_00_CreateSecretTable.sql [hunk 2]
```diff
@@ -0,0 +1,22 @@
+IF OBJECT_ID('[dbo].[Secret]') IS NULL
+BEGIN
+	CREATE TABLE [dbo].[Secret] (
+		[Id] UNIQUEIDENTIFIER NOT NULL,
+		[OrganizationId] UNIQUEIDENTIFIER NOT NULL,
+        [Key] NVARCHAR(MAX) NULL,
```
Comment (Hinton) [inline util/Migrator/DbScripts/2022-07-20_00_CreateSecretTable.sql:6]: We seem to be mixing spaces and tabs in this file. Please ensure it uses spaces.

**Review by Hinton — CHANGES_REQUESTED** [review]
Looks good overall.

A couple of comments we should look into.

I will also need to look into the tests more in-depth. I know most EF test were build to compare dapper and EF, this is not the case for secrets.

File: src/Core/Entities/Secret.cs [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+        public void SetNewId()
+        {
+            if (Id == default(Guid))
+            {
+                Id = CoreHelpers.GenerateComb();
+            }
+        }
+
+    }
```
[the remark this answers is not in the document]
Comment (Thomas-Avery) [reply src/Core/Entities/Secret.cs:33]: This is needed for Create implementation in Repository.cs

File: src/Infrastructure.EntityFramework/Repositories/DatabaseContext.cs (hunk 1 above)
Comment (Thomas-Avery) [reply src/Infrastructure.EntityFramework/Repositories/DatabaseContext.cs]: I think using EntityTypeConfiguration files is a good idea and will be simple to add more configurations in the future.

Should now be updated in the latest commit.

File: src/Infrastructure.EntityFramework/Repositories/DatabaseContext.cs [hunk 4]
```diff
@@ -141,7 +135,9 @@ protected override void OnModelCreating(ModelBuilder builder)
             eUser.ToTable(nameof(User));
             eOrganizationApiKey.ToTable(nameof(OrganizationApiKey));
             eOrganizationConnection.ToTable(nameof(OrganizationConnection));
-            eSecret.ToTable(nameof(Secret));
+
+            // Apply all configurations specified in types implementing IEntityTypeConfiguration in the assembly.
+            builder.ApplyConfigurationsFromAssembly(typeof(SecretEntityTypeConfiguration).Assembly);
```
Comment (Hinton) [inline src/Infrastructure.EntityFramework/Repositories/DatabaseContext.cs:140]: I assume this is to get the assembly of the Infrastructure.EF project? Should we use the `DatabaseContext` file instead which should have the same effect but is less coupled to a single configuration.

File: src/Infrastructure.EntityFramework/Repositories/DatabaseContext.cs [hunk 5]
```diff
@@ -141,7 +135,9 @@ protected override void OnModelCreating(ModelBuilder builder)
             eUser.ToTable(nameof(User));
             eOrganizationApiKey.ToTable(nameof(OrganizationApiKey));
             eOrganizationConnection.ToTable(nameof(OrganizationConnection));
-            eSecret.ToTable(nameof(Secret));
+
+            // Apply all configurations specified in types implementing IEntityTypeConfiguration in the assembly.
```
Comment (Hinton) [inline src/Infrastructure.EntityFramework/Repositories/DatabaseContext.cs:139]: This comment is slightly reduncant, lets try and make it more generic.

I believe we should move this to the top of this method, and add a note that the other way is deprecated.

```c#
// Scans and loads all configurations implementing the `IEntityTypeConfiguration` from the
//  `Infrastructure.EntityFramework` Module. Note to get the assembly we can use a random class
//   from this module.
```

File: src/Infrastructure.EntityFramework/Configurations/SecretEntityTypeConfiguration.cs [hunk 6]
```diff
@@ -0,0 +1,27 @@
+﻿using Bit.Infrastructure.EntityFramework.Models;
+using Microsoft.EntityFrameworkCore;
+using Microsoft.EntityFrameworkCore.Metadata.Builders;
+
+public class SecretEntityTypeConfiguration : IEntityTypeConfiguration<Secret>
+{
+    public void Configure(EntityTypeBuilder<Secret> builder)
+    {
+        builder
+            .Property(b => b.Id)
```
Comment (Hinton) [inline src/Infrastructure.EntityFramework/Configurations/SecretEntityTypeConfiguration.cs:10]:
```suggestion
            .Property(s => b.Id)
```

File: util/SqlServerEFScaffold/SqlServerEFScaffold.csproj [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+      <PrivateAssets>all</PrivateAssets>
+    </PackageReference>
+  </ItemGroup>
+
+  <ItemGroup>
+    <Compile Include="..\EfShared\MigrationBuilderExtensions.cs" />
+  </ItemGroup>
+
+  <ItemGroup>
+    <ProjectReference Include="..\..\src\Api\Api.csproj" />
```
Comment (Hinton) [inline util/SqlServerEFScaffold/SqlServerEFScaffold.csproj:14]: Do we need to reference Api?

**Review by Hinton — CHANGES_REQUESTED** [review]
We need to add the database table to `Sql/dbo/tables` since that folder is expected to be an accurate representation of the current database model.

File: util/PostgresMigrations/Factories.cs [hunk 8]
```diff
@@ -24,6 +24,7 @@ public DatabaseContext CreateDbContext(string[] args)
             var globalSettings = GlobalSettingsFactory.GlobalSettings;
             var optionsBuilder = new DbContextOptionsBuilder<DatabaseContext>();
             var connectionString = globalSettings.PostgreSql?.ConnectionString;
+            AppContext.SetSwitch("Npgsql.EnableLegacyTimestampBehavior", true);
```
[the remark this answers is not in the document]
Comment (Hinton) [reply util/PostgresMigrations/Factories.cs:28]: Please expand the comment to explain why we enable it as the code already describes what we do.

```
NpgSql 6.0 changed how timezones works, and migrated to use `timestamp with time zone`. We have not yet updated our projects to use this format and need to fallback to the previous behavior. Check https://www.npgsql.org/doc/release-notes/6.0.html#timestamp-rationalization-and-improvements for more details.

File: util/SqlServerEFScaffold/SqlServerEFScaffold.csproj (hunk 7 above)
Comment (Thomas-Avery) [reply util/SqlServerEFScaffold/SqlServerEFScaffold.csproj:14]: Yes for this line
```            var configBuilder = new ConfigurationBuilder().AddUserSecrets<Bit.Api.Startup>();```

File: src/Infrastructure.EntityFramework/Configurations/SecretEntityTypeConfiguration.cs (hunk 6 above)
Comment (Thomas-Avery) [reply src/Infrastructure.EntityFramework/Configurations/SecretEntityTypeConfiguration.cs]: resolved in https://github.com/bitwarden/server/pull/2144/commits/9c410672d8b381974f1ab20a384028eb21fab8ed

File: src/Infrastructure.EntityFramework/Repositories/DatabaseContext.cs (hunk 5 above)
Comment (Thomas-Avery) [reply src/Infrastructure.EntityFramework/Repositories/DatabaseContext.cs]: resolved in https://github.com/bitwarden/server/pull/2144/commits/9c410672d8b381974f1ab20a384028eb21fab8ed

File: src/Infrastructure.EntityFramework/Repositories/DatabaseContext.cs (hunk 4 above)
Comment (Thomas-Avery) [reply src/Infrastructure.EntityFramework/Repositories/DatabaseContext.cs]: Resolved in https://github.com/bitwarden/server/pull/2144/commits/9c410672d8b381974f1ab20a384028eb21fab8ed

File: src/Core/Repositories/ISecretRepository.cs [hunk 9]
```diff
@@ -4,8 +4,8 @@ namespace Bit.Core.Repositories
 {
     public interface ISecretRepository
     {
-        Task<IEnumerable<Secret>> GetManyByOrganizationIdAsync(Guid organizationId);
-        Task<Secret> GetByIdAsync(Guid id);
+        Task<IEnumerable<Secret>> GetManyByOrganizationIdAsync(Guid organizationId, bool includeDeleted = false);
+        Task<Secret> GetByIdAsync(Guid id, bool includeDeleted = false);
```
Comment (Hinton) [inline src/Core/Repositories/ISecretRepository.cs:8]: I suspect we would want this to behave slightly different. You most likely want to fetch a list of only trash items, or only non trash items.

Comment (Thomas-Avery) [reply src/Core/Repositories/ISecretRepository.cs]: resolved in https://github.com/bitwarden/server/pull/2144/commits/3481141d596c226572e6b0d893e70d716e0658c1

**Comment by Hinton** [comment]
Looks good, please resolve the last comment.

File: util/PostgresMigrations/Factories.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (Thomas-Avery) [reply util/PostgresMigrations/Factories.cs:29]: resolved in https://github.com/bitwarden/server/pull/2144/commits/7f36e8fd1700806d2c6b5513304dbec6a821747a

File: bitwarden_license/src/CommCore/CommCore.csproj [hunk 10]
```diff
@@ -2,6 +2,7 @@
 
   <ItemGroup>
     <ProjectReference Include="..\..\..\src\Core\Core.csproj" />
+    <ProjectReference Include="..\Commercial.Infrastructure.EntityFramework\Commercial.Infrastructure.EntityFramework.csproj" />
```
Comment (Hinton) [inline bitwarden_license/src/CommCore/CommCore.csproj:5]: Core shouldn't depend on infrastructure.

Comment (Thomas-Avery) [reply bitwarden_license/src/CommCore/CommCore.csproj]: resolved in https://github.com/bitwarden/server/pull/2144/commits/5d6a2145bcf5210ab3c7a59a440cfa06aa87f6d0

File: bitwarden_license/src/Commercial.Core/packages.lock.json [hunk 11]
```diff
@@ -2441,6 +2603,15 @@
...
       },
+      "commercial.infrastructure.entityframework": {
+        "type": "Project",
+        "dependencies": {
+          "AutoMapper.Extensions.Microsoft.DependencyInjection": "11.0.0",
+          "Core": "2022.6.2",
+          "Infrastructure.EntityFramework": "2022.6.2",
+          "Microsoft.EntityFrameworkCore": "6.0.4"
+        }
+      },
```
Comment (Hinton) [inline bitwarden_license/src/Commercial.Core/packages.lock.json:2614]: I think we need to re-generate the lock file. Core shouldn't depend on EF

**Review by Hinton — APPROVED** [review]
Looks good!

**Pull request merged**